[spec] [footnotes] Convert footnotes to asciidoc

hussainnm opened the pull request:
* Added footnotes to all chapters
 (except EnterpriseBeanEnvironment.adoc fixed in #74 )
* Removed Footnotes.adoc from EBCore.adoc

kwsutter commented on spec/src/main/asciidoc/core/ExceptionHandling.adoc:
```diff
@@ -984,13 +1031,19 @@ instance).
 If the exception or error happened during the
 processing of a client invoked method, throw the
-_javax.ejb.EJBExceptionlink:#a10308[88]_ . If the EJB 2.1 client
+_javax.ejb.EJBException_footnote:a10308[If the business interface is 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/ExceptionHandling.adoc:
```diff
@@ -984,13 +1031,19 @@ instance).
 local client. If the instance executed in the client’s transaction, the
 container should throw the
-_javax.ejb.EJBTransactionRolledbackExceptionlink:#a10309[89]_ .
+_javax.ejb.EJBTransactionRolledbackException_footnote:a10309[If the 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/ExceptionHandling.adoc:
```diff
@@ -1002,7 +1055,10 @@ that it is fruitless to continue the transaction.)
 If a client makes a call to a stateful
 session object that has been removed, the container should throw the
-_javax.ejb.NoSuchEJBExceptionlink:#a10310[90]_ . If the EJB 2.1
+_javax.ejb.NoSuchEJBException_footnote:a10310[If the business interface 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/ExceptionHandling.adoc:
```diff
@@ -1019,7 +1075,10 @@ that may be thrown by the transaction start and commit operations.
 If the container fails to start or commit a
 container-managed transaction, the container must throw the
-_javax.ejb.EJBExceptionlink:#a10311[91]_ . If the web service
+_javax.ejb.EJBException_footnote:a10311[If the business interface is a 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/Interceptors.adoc:
```diff
@@ -164,7 +165,7 @@ _InvocationContext.proceed()_ is called.
 chain, the bean instance and any associated interceptor instances are
 discarded (unless the bean is a singleton session
-beanlink:#a10262[43]). The _PreDestroy_ callbacks are not invoked
+beanfootnote:a10262[See <<a1254>>.]). The _PreDestroy_ callbacks are not invoked
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/Overview.adoc:
```diff
@@ -80,7 +80,11 @@ perform the two EJB Roles of the Enterprise Bean Provider and the
 
 The following sections define the six EJB
-roleslink:#a10218[1].
+rolesfootnote:a10218[Earlier releases of this specification 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/MessageDrivenBeanComponentContract.adoc:
```diff
@@ -784,7 +786,11 @@ of a message-driven bean instance:
 A message-driven bean instance’s life starts
 when the container invokes the message-driven bean class constructor to
-create a new instancelink:#a10259[40]. Next, the container
+create a new instancefootnote:a10259[If an `AroundConstruct` lifecycle 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/MessageDrivenBeanComponentContract.adoc:
```diff
@@ -299,7 +299,9 @@ supported for message-driven beans.
 The PostConstruct and PreDestroy callback
 methods may be defined directly on the bean class or on a separate
-interceptor classlink:#a10258[39]. The AroundConstruct lifecycle
+interceptor classfootnote:a10258[If `PrePassivate` or `PostActivate` 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/SecurityManagement.adoc:
```diff
@@ -702,7 +702,9 @@ application or its deployment descriptor refers to a role named “**”.
 Defining the security roles in the metadata
 annotations and/or deployment descriptor is
-optionallink:#a10327[107]. Their omission means that the Bean
+optionalfootnote:a10327[If the Bean Provider and Application Assembler do 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/SecurityManagement.adoc:
```diff
@@ -838,7 +840,9 @@ Method permissions are defined as a binary
 business interfaces, home interfaces, component interfaces, no-interface
 views, and/or web service endpoints of session and
-entitylink:#a10328[108] beans, including all their
+entityfootnote:a10328[Component contract and client view of entity beans 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/SecurityManagement.adoc:
```diff
@@ -1630,7 +1634,9 @@ The EJB container must be capable of allowing
 execution of the business, home, and component interfaces, no-interface
 view, and/or web service endpoint methods of a session or an
-entitylink:#a10329[109] bean, or for the message listener methods
+entityfootnote:a10329[Component contract and client view of entity beans 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/SecurityManagement.adoc:
```diff
@@ -1713,7 +1719,10 @@ security roles that includes the method in its method permissions
 _all_ the roles associated with the method.) If the container denies a
 client access to a business method, the container should throw the
-_javax.ejb.EJBAccessExceptionlink:#a10330[110]_ . If the EJB 2.1
+_javax.ejb.EJBAccessException_footnote:a10330[If the business interface 
```
Need a separator before `footnote`.

kwsutter commented on spec/src/main/asciidoc/core/SecurityManagement.adoc:
```diff
@@ -1749,7 +1758,9 @@ managers according to the security policy defined by the Deployer.
 The container must allow the same enterprise
 bean to be deployed independently multiple times, each time with a
-different security policylink:#a10331[111]. The container must
+different security policyfootnote:a10331[For example, the enterprise 
```
Need a separator before `footnote`.

hussainnm commented:
I will work on these comments.

hussainnm commented on spec/src/main/asciidoc/core/ExceptionHandling.adoc:
```diff
@@ -430,7 +450,9 @@ Log the exception or error.
 Discard instance.
 
 {empty}Throw _EJBException_ to
-client.link:#a10292[73]
```
"client." is removed. Put it back.

hussainnm commented:
I moved the footnotes after the punctuation marks if it is present, otherwise I have added {empty} attribute to separate the word from the footnote. This still does not resolve your concern on making the asciidoc readable. Adding a space or new line to start the footnote will also create a space between the word and footnote in the rendered document.

One other issue I noticed in the following lines
* SecurityManagement.adoc#843,#1637
* SupportForDistributedInteroperability.adoc#62,#105

The footnote appears in between entity and beans.

See: Section 10.2, 10.2.1, 12.3.2, 12.6.4.

Should I move the footnote after beans?

kwsutter commented on spec/src/main/asciidoc/core/CompatibilityAndMigration.adoc:
```diff
@@ -11,7 +11,7 @@ clients.
 Existing EJB 3.1 and earlier applications
 that do not include entity bean components and JAX-RPC based web
-services endpointsfootnote:a10344[See <<a182>>.] must be supported to run
+services endpoints{empty}footnote:a10344[See <<a182>>.] must be supported to run
```
This comment hits several files.  Let's just figure out an approach here and then apply it across the others...

kwsutter commented:
> One other issue I noticed in the following lines
> 
>     * SecurityManagement.adoc#843,#1637
> 
>     * SupportForDistributedInteroperability.adoc#62,#105
> 
> 
> The footnote appears in between entity and beans.
> 
> See: Section 10.2, 10.2.1, 12.3.2, 12.6.4.
> 
> Should I move the footnote after beans?

Same comment as my review above...  Inserting a space makes everything still readable, both in source and rendered formats.  The footnote superscript [1] does not have to be flush with the word.  As an example, I would do ".. entity footnote:a1234[footnote text] beans.."

hussainnm commented:
* Replaced {empty} with single space before footnote
* Added single space before footnote after punctuation

Re-based on master and made above changes to Chapter 11.

tkburroughs commented:
This is looking good; too bad asciidoc does't support footnotes at the bottom of the page.

Perhaps I missed it, but in the latest changes, why "Added single space before footnote after punctuation"?

For example, footnote 2 seems odd because there is a space after the "."... but no space between the footnote and the next sentence.  This makes it appear that the footnote is associated with the next sentence, not the one that just ended.  And for some reason, the PDF doc places the closing ] on the following line with the first word of the next sentence.   Seems we really don't want the space after the punctuation, but we do want a space before the next sentence.

hussainnm commented:
I will make the changes such that if the footnote comes at the end of the sentence then put the footnote before the punctuation. For e.g., _"by reference" footnote:a10219[...]_ will be rendered as _"by reference" [2]._

If the footnote is in between a sentence then I will put a space between the footnote and both the words. For e.g., _endpoints footnote:a10344[...] must_ will be rendered as _endpoints [124] must_

hussainnm commented:
@tkburroughs I have made the requested changes. Please review.

tkburroughs commented:
@hussainnm approved and merged
